### PR TITLE
Response priority management in admin

### PR DESCRIPTION
**Pull request by davea**
This PR upgrades the hard-coded `Oxfordshire::problem_response_priorities` method into a fully-fledged `ResponsePriority` model which can be managed via the admin. Priorities are assigned to bodies and categories, and Problems can optionally have a priority set.

For mysociety/fixmystreetforcouncils#66

### CODE REVIEW

File: db/schema_0045-response-priorities.sql [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+);
+
+CREATE TABLE contact_response_priorities (
+    id serial NOT NULL PRIMARY KEY,
+    contact_id int REFERENCES contacts(id) NOT NULL,
+    response_priority_id int REFERENCES response_priorities(id) NOT NULL
+);
+
+ALTER TABLE problem
+    ADD COLUMN response_priority_id int REFERENCES response_priorities(id);
```
Comment (dracos) [inline db/schema_0045-response-priorities.sql]: You've probably already spotted this, this line needs adding in schema.sql.

File: perllib/FixMyStreet/App/Controller/Admin/ResponsePriorities.pm [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    my ( $self, $c ) = @_;
+
+    $c->forward('/admin/begin');
+}
+
+sub auto : Private {
+    my ( $self, $c ) = @_;
+
+    $c->forward('/admin/auto');
+}
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Admin/ResponsePriorities.pm]: All `auto` functions in the chain are called (unlike `begin` or `end`) so this function shouldn't be needed, if it is, something must be going wrong somewhere :)

File: perllib/FixMyStreet/App/Controller/Admin/ResponsePriorities.pm [hunk 3]
```diff
@@ -0,0 +1,125 @@
+package FixMyStreet::App::Controller::Admin::ResponsePriorities;
+use Moose;
+use namespace::autoclean;
+
+BEGIN { extends 'Catalyst::Controller'; }
+
+use Path::Class;
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Admin/ResponsePriorities.pm]: Not sure any of these are used.

File: perllib/FixMyStreet/App/Controller/Admin/ResponsePriorities.pm [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    my $has_permission = $c->user->has_body_permission_to('responsepriority_edit') &&
+                         $c->user->from_body->id eq $body_id;
+
+    unless ( $c->user->is_superuser || $has_permission ) {
+        $c->detach( '/page_error_404_not_found' );
+    }
+
+    # Regular users can only view their own body's priorities
+    if ( !$c->user->is_superuser && $body_id ne $c->user->from_body->id ) {
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Admin/ResponsePriorities.pm]: Don't think you can ever get to this if passing, due to the 404 check above. Don't mind whether it 404s or redirects, 404 probably easier :)

File: db/schema_0045-response-priorities.sql (hunk 1 above)
Comment (dracos) [inline db/schema_0045-response-priorities.sql]: General query - could priority have stayed as extra metadata?

Comment (davea) [inline db/schema_0045-response-priorities.sql]: I thought about it, and similar to `users.area_id` thought it's likely we'll want to index/filter on this field in the near future.

**Comment by dracos** [comment]
Would be good to have some tests of the admin pages. Tests for the front end inspect page showing/accepting the right things probably easier after merging this and then #1516.
🍨

**Pull request merged**